src/rpc/ngHandlers/NFTBuyOffers.cpp

```diff
@@ -0,0 +1,226 @@
+
+NFTBuyOffersHandler::Result
+NFTBuyOffersHandler::process(
+    NFTBuyOffersHandler::Input input,
+    boost::asio::yield_context& yield) const
+{
+    auto const tokenID = ripple::uint256{input.nftID.c_str()};
+    auto const range = sharedPtrBackend_->fetchLedgerRange();
+    auto const lgrInfoOrStatus = getLedgerInfoFromHashOrSeq(
+        *sharedPtrBackend_,
+        yield,
+        input.ledgerHash,
+        input.ledgerIndex,
+        range->maxSequence);
+    if (auto const status = std::get_if<Status>(&lgrInfoOrStatus))
+        return Error{*status};
+
+    auto const lgrInfo = std::get<LedgerInfo>(lgrInfoOrStatus);
+    auto const directory = keylet::nft_buys(tokenID);
+
+    // TODO: just check for existence without pulling
+    if (not sharedPtrBackend_->fetchLedgerObject(
+            directory.key, lgrInfo.seq, yield))
+        return Error{Status{RippledError::rpcOBJECT_NOT_FOUND, "notFound"}};
+
+    auto output = NFTBuyOffersHandler::Output{input.nftID};
+    auto offers = std::vector<ripple::SLE>{};
+    auto reserve = input.limit;
+    auto cursor = uint256{};
+    auto startHint = uint64_t{0ul};
+
+    if (input.marker)
+    {
+        cursor = uint256(input.marker->c_str());
+
+        // We have a start point. Use limit - 1 from the result and use the very
+        // last one for the resume.
+        auto const sle =
+            [this, &cursor, &lgrInfo, &yield]() -> std::shared_ptr<SLE const> {
+            auto key = keylet::nftoffer(cursor).key;
+            if (auto const blob = sharedPtrBackend_->fetchLedgerObject(
+                    key, lgrInfo.seq, yield);
+                blob)
+            {
+                return std::make_shared<SLE const>(
+                    SerialIter{blob->data(), blob->size()}, key);
+            }
+            return nullptr;
+        }();
+
+        if (!sle ||
+            sle->getFieldU16(ripple::sfLedgerEntryType) !=
+                ripple::ltNFTOKEN_OFFER ||
+            tokenID != sle->getFieldH256(ripple::sfNFTokenID))
+        {
+            return Error{Status{RippledError::rpcINVALID_PARAMS}};
+        }
+
+        startHint = sle->getFieldU64(ripple::sfNFTokenOfferNode);
+        output.offers.push_back(*sle);
+        offers.reserve(reserve);
+    }
+    else
+    {
+        // We have no start point, limit should be one higher than requested.
+        offers.reserve(++reserve);
+    }
+
+    auto result = traverseOwnedNodes(
+        *sharedPtrBackend_,
+        directory,
+        cursor,
+        startHint,
+        lgrInfo.seq,
+        reserve,
+        {},
+        yield,
+        [&offers](ripple::SLE&& offer) {
+            if (offer.getType() == ripple::ltNFTOKEN_OFFER)
+            {
+                offers.push_back(std::move(offer));
+                return true;
+            }
+
+            return false;
+        });
+
+    if (auto status = std::get_if<Status>(&result))
+        return Error{*status};
+
+    if (offers.size() == reserve)
```

Comment:
If there are other types in this directory, the size might be less than reserve, but we still need to include the marker.

Reply:
Good point. I copied the code tho so not sure if this is a bug or not. We can deal with it after all rpc is migrated and tested i think.